### PR TITLE
Filter consolidation

### PR DESCRIPTION
**Pull request by richeduni**

### CODE REVIEW

File: components/MapFilters.vue [hunk 1]
```diff
@@ -74,171 +107,204 @@
...
+           let witch = marker.witches[w];
 
-         if (witch[this.currentFilterProperty] !== filterType) {
-           filteredWitches.push(witch);
+           if (witch[filterProperty] === filterType) {
+             witch.witchState.onOff = false;
+             witch.witchState.activeFilters.push(filterProperty);
+           }
          }
+         [marker.markerIcon, marker.onOff] = this.getMarkerState(marker);
```
Comment (richeduni) [inline components/MapFilters.vue:196]: Please add new line above [marker.markerIcon, marker.onOff] = this.getMarkerState(marker);

For control structures, our team add a new line above and below for readability. e.g.

```
let x =1;
for(let i = 0; i < 10; i++) {  
    console.log(i);
}
let y = 5;
```

Should be:

```
let x =1;

for(let i = 0; i < 10; i++) {  
    console.log(i);
}

let y = 5;

```
Same for if statements

File: components/MapFilters.vue [hunk 2]
```diff
@@ -74,171 +107,204 @@
...
      },
-     getWitchesFromArray: function (witches, filterType) {
-       // Given an array of witches, it returns the an array
-       // with the ones that meet the property filterType.
-       let filteredWitches = [];
+     getUpdatedWitchFilters (activeFilters, filterProperty) {
+       if (activeFilters.length === 1) {
+         return [];
+       }
+       return activeFilters.filter(function (activeProperty) {
```
Comment (richeduni) [inline components/MapFilters.vue:203]: Please add new line above return statement

File: components/MapFilters.vue [hunk 3]
```diff
@@ -74,171 +107,204 @@
...
+     },
+     setWitchesOn: function (filterProperty, filterType) {
+       // Filters <filterType> on. It goes through the current markers
+       // setting witches that meet filter type to on and updating the witches
+       // active filters. If a marker is not mixed, meaning it could become
+       // mixed, it updates the marker state by calling getMarkerState.
 
-       for (let i = 0, len = witches.length; i < len; i++) {
-         let witch = witches[i]
+       for (let i = 0, len = this.markers.length; i < len; i++) {
```
Comment (richeduni) [inline components/MapFilters.vue:213]: Please use standard for loop

File: components/MapFilters.vue [hunk 4]
```diff
@@ -74,171 +107,204 @@
...
+           if (witchType === filterType) {
+             let newFilters = this.getUpdatedWitchFilters(witch.witchState.activeFilters, filterProperty);
+             witch.witchState.activeFilters = newFilters;
+
+             if (newFilters.length === 0) {
+               witch.witchState.onOff = true;
+             }
+           }
          }
+         [marker.markerIcon, marker.onOff] = this.getMarkerState(marker);
```
Comment (richeduni) [inline components/MapFilters.vue:229]: Please add new line below for loop for readability

File: components/MapFilters.vue [hunk 5]
```diff
@@ -74,171 +107,204 @@
...
+       // witches that are on. This is the array that will be
+       // emitted to parent, and will then be used by LeafletMap
+       // to plot.
+       let outputMarkers = [];
 
-       this.markers.forEach(marker => {
-         marker.witches = this.removeWitchesFromArray(marker.witches, filterType);
-         this.updateMarkerState(marker);
-       });
+       for (let i = 0, len = this.markers.length; i < len; i++) {
```
Comment (richeduni) [inline components/MapFilters.vue:247]: Please use standard for loop

File: components/MapFilters.vue [hunk 6]
```diff
@@ -74,171 +107,204 @@
...
 
-       this.noFiltersOff += 1; // Keeping track of the number of inactive filters.
+         if (marker.onOff) {
+           let activeWitches = marker.witches.filter(function (witch) {
+             return witch.witchState.onOff;
+           });
+           outputMarkers.push(this.buildOutputMarker(marker, activeWitches));
+         }
+       }
+       return outputMarkers;
```
Comment (richeduni) [inline components/MapFilters.vue:257]: Please add new line above return statement

File: components/MapFilters.vue [hunk 7]
```diff
@@ -74,171 +107,204 @@
...
+     setFilterInactive: function (property, filterType) {
+       this.filterProperties[property].filters[filterType].active = false;
      },
-     filterMarkers: function (filterType) {
-       let isActive = this.filterLayers[this.currentLayer].filters[filterType].active;
+     setFilterActive: function (property, filterType) {
+       this.filterProperties[property].filters[filterType].active = true;
+     },
+     filterMarkers: function (property, filterType) {
+       console.log(property);
```
Comment (richeduni) [inline components/MapFilters.vue:266]: Please remove console.log statement

File: components/MapFilters.vue [hunk 8]
```diff
@@ -74,171 +107,204 @@
...
        }
-       else {
-         this.filterOff(filterType);
+     },
+     setAllIcons: function () {
+       // Goes through all markers changing the icons according
+       // to the current filter property. Called in togglefilterProperties
+       // when user changes filter properties so that icons update
+       // accordingly.
+       for (let i = 0, len = this.markers.length; i < len; i++) {
```
Comment (richeduni) [inline components/MapFilters.vue:284]: Please use standard for loop

File: components/MapFilters.vue [hunk 9]
```diff
@@ -74,171 +107,204 @@
...
-       this.currentLayer = layerIndex;
-       this.setAllMarkerIcons()
-       this.$emit("updatedMarkers", this.markers);
+     setPropertyToShowing: function (property) {
+       // Sets the property <property> to showing, and calls
+       // setPropertyToCurrent to set the property as the current
+       // current property, and change the icons accordingly.
+       if (this.currentProperty !== property) {
+         this.setPropertyToCurrent(property);
+       }
```
Comment (richeduni) [inline components/MapFilters.vue:302]: Please add new line after if statement

File: components/MapFilters.vue [hunk 10]
```diff
@@ -248,14 +314,63 @@
        this.filters = ! this.filters;
      }
    },
-   computed: {
-     currentFilterProperty: function() {
-       return this.filterLayers[this.currentLayer].property;
-     }
+   mounted: function () {
+     console.log(this.filterProperties);
```
Comment (richeduni) [inline components/MapFilters.vue:318]: Please remove console.log call

File: components/MapFilters.vue [hunk 11]
```diff
@@ -248,14 +314,63 @@
        this.filters = ! this.filters;
```
Comment (richeduni) [inline components/MapFilters.vue:314]: toggleFilters is never used. please remove

File: pages/death.vue [hunk 12]
```diff
@@ -227,19 +186,18 @@
...
-           // add to social class filter if doesn't exist already.
-           let socialsFound = Object.keys(this.filterLayers[1].filters);
+           let newSocial = APIDataHandler.checkFilters(socialClassification, currentSocials, icons);
+           let newOccupation = APIDataHandler.checkFilters(occupation, currentOccupations, icons);
 
-           if(!socialsFound.find(socialFound => socialFound === socialClassification)){
-             this.filterLayers[1].filters[socialClassification] = {label:socialClassification, active: true, iconUrl: this.icons[socialsFound.length]};
+           if (newSocial) {
+             this.filterProperties.socialClass.filters[newSocial.label] = newSocial;
            }
```
Comment (richeduni) [inline pages/death.vue:198]: Please add new line after if statement

File: pages/death.vue [hunk 13]
```diff
@@ -251,7 +209,6 @@
            // push
            if (witch) {
              if (id === "http://www.wikidata.org/entity/Q43394934") {
```
Comment (richeduni) [inline pages/death.vue:211]: Please remove, as this doesn't do anything

File: pages/death.vue [hunk 14]
```diff
@@ -316,24 +277,24 @@
        });
 
        // if a marker exists for the witche's location add the witch to it. if not create a new marker for the location and add the witch.
-       let startingProperty = this.filterLayers[this.startingLayer].property;
+       let filterProperty = 'sex';
 
        if (marker) {
          marker.witches.push(witch);
 
          for (let i = 0, len = marker.witches.length; i < len; i++) {
```
Comment (richeduni) [inline pages/death.vue:285]: Please use standard for loop

File: pages/detention.vue [hunk 15]
```diff
@@ -228,18 +188,18 @@
...
-           // add to occupations filters if doesn't exist already.
-           let occupationsFound = Object.keys(this.filterLayers[2].filters);
+           let newSocial = APIDataHandler.checkFilters(socialClassification, currentSocials, icons);
+           let newOccupation = APIDataHandler.checkFilters(occupation, currentOccupations, icons);
 
-           if(!occupationsFound.find(occupationFound => occupationFound === occupation)){
-             this.filterLayers[2].filters[occupation] = {label:occupation, active: true, iconUrl: this.icons[occupationsFound.length]};
+           if (newSocial) {
+             this.filterProperties.socialClass.filters[newSocial.label] = newSocial;
+           }
```
Comment (richeduni) [inline pages/detention.vue:200]: Please add new line after if statement

File: pages/index.vue [hunk 16]
```diff
@@ -239,20 +197,19 @@
...
+           let newOccupation = APIDataHandler.checkFilters(occupation, currentOccupations, icons);
 
-           // add to social class filter if doesn't exist already.
-           let socialsFound = Object.keys(this.filterLayers[1].filters);
-           
-           if(!socialsFound.find(socialFound => socialFound === socialClassification)){
-             this.filterLayers[1].filters[socialClassification] = {label:socialClassification, active: true, iconUrl: this.icons[socialsFound.length]};
+           if (newSocial) { 
+             this.filterProperties.socialClass.filters[newSocial.label] = newSocial;
            }
```
Comment (richeduni) [inline pages/index.vue:210]: Please add new line after if statement

File: pages/index.vue [hunk 17]
```diff
@@ -323,36 +284,37 @@
...
 
        // if a marker exists for the witche's location add the witch to it. if not create a new marker for the location and add the witch.
-       let startingProperty = this.filterLayers[this.startingLayer].property;
-
+       let filterProperty = 'sex';
+      
        if (marker) {
          marker.witches.push(witch);
 
          for (let i = 0, len = marker.witches.length; i < len; i++) {
```
Comment (richeduni) [inline pages/index.vue:305]: Please use standard for loop

File: pages/death.vue (hunk 14 above)
Comment (josep-reyero) [reply pages/death.vue:285]: I thought this is a standard for loop?

File: components/MapFilters.vue [hunk 18]
```diff
@@ -74,171 +107,204 @@
...
-       type: Boolean,
+     startingFilters: {
+       type: Object,
        required: true
      }
    },
    data () {
      return {
        filters: false,
        noFiltersOn: 0,
```
[the remark this answers is not in the document]
Comment (josep-reyero) [reply components/MapFilters.vue]: Removed.

File: components/MapFilters.vue [hunk 19]
```diff
@@ -74,171 +107,204 @@
...
-       // one witch has a different type to the others,
-       // returns mixed straight away. Otherwise returns
-       // markerType, which will be common to all witches.
+     getMarkerIcon: function (markerType) {
+       // Returns the marker icon based on markerType (markerType
+       // is eithera filterType or 'mixed').
 
-       let markerType = '';
-       let witches = marker.witches;
+       if (markerType === 'mixed') {
```
[the remark this answers is not in the document]
Comment (josep-reyero) [reply components/MapFilters.vue:162]: Resolved - I think.

File: pages/death.vue (hunk 13 above)
Comment (josep-reyero) [reply pages/death.vue]: Resolved.

File: pages/detention.vue (hunk 15 above)
Comment (josep-reyero) [reply pages/detention.vue:194]: Resolved.

File: pages/index.vue (hunk 16 above)
Comment (josep-reyero) [reply pages/index.vue:207]: Resolved.

File: pages/index.vue (hunk 17 above)
Comment (josep-reyero) [reply pages/index.vue]: Resolved.

File: assets/js/APIDataHandler.js [hunk 20]
```diff
@@ -0,0 +1,27 @@
+const APIDataHandler = {
```
[the remark this answers is not in the document]
Comment (josep-reyero) [reply assets/js/APIDataHandler.js]: Resolved.

File: assets/js/APIDataHandler.js [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+    if (!propertyFilters.includes(filter)) {
+      let icon = this.selectIcon(propertyFilters, icons);
+      
+      return {
+        label: filter,
+        active: true,
+        iconUrl: icon
+      }
+    }
+    return null
```
[the remark this answers is not in the document]
Comment (josep-reyero) [reply assets/js/APIDataHandler.js]: Resolved.

File: components/MapFilters.vue [hunk 22]
```diff
@@ -74,171 +107,204 @@
...
+       }
+       return this.filterProperties[this.currentProperty].filters[markerType].iconUrl;
+     },
+     getMarkerType: function (witches) {
+       // Gets the marker type (either a filterType, 'mixed' or null if
+       // no witches are on) based on those witches that are on. Used
+       // to get the new marker type after setting witches on/off.
+       let markerType = null;
 
        for (let i = 0, len = witches.length; i < len; i++) {
```
[the remark this answers is not in the document]
Comment (josep-reyero) [reply components/MapFilters.vue:174]: Resolved.

File: components/MapFilters.vue [hunk 23]
```diff
@@ -74,171 +107,204 @@
...
+
+           if (!markerType) {
+             markerType = witchType;
+           } else if (witchType !== markerType) {
+             return 'mixed';
+           }
          }
        }
-
        return markerType;
```
[the remark this answers is not in the document]
Comment (josep-reyero) [reply components/MapFilters.vue:187]: resolved.

File: components/MapFilters.vue [hunk 24]
```diff
@@ -74,171 +107,204 @@
...
-       let markerType = this.getMarkerType(marker);
-
-       if (markerType === 'mixed') {
-         return '/images/witch-single-purple.png';
-       } else {
-         return this.filterLayers[this.currentLayer].filters[markerType].iconUrl;
+       if (markerType) {
+         return [this.getMarkerIcon(markerType), true];
        }
+       return [null, false];
```
[the remark this answers is not in the document]
Comment (josep-reyero) [reply components/MapFilters.vue:200]: Resolved.

File: components/MapFilters.vue [hunk 25]
```diff
@@ -74,171 +107,204 @@
...
-       // filtered. Specifically, its marker Icon and
-       // if it needs to be on or off depending on whether
-       // all the witches have been filtered out.
-
-       if (marker.witches.length > 0) {
-         marker.markerIcon = this.getMarkerIcon(marker);
-         marker.onOff = true;
-       } else {
-         marker.onOff = false;
+     getIsMixed: function (marker) {
```
[the remark this answers is not in the document]
Comment (josep-reyero) [reply components/MapFilters.vue:202]: Resolved.

File: components/MapFilters.vue [hunk 26]
```diff
@@ -74,171 +107,204 @@
...
+       // If a marker is mixed, meaning it could stop being mixed, it
+       // updates the marker state by calling getMarkerState.
 
-       let filteredWitches = [];
+       for (let i = 0, len = this.markers.length; i < len; i++) {
+         let marker = this.markers[i];
 
-       for (let i = 0, len = witches.length; i < len; i++) {
-         let witch = witches[i]
+         for (let w = 0, len = marker.witches.length; w < len; w++) {
```
[the remark this answers is not in the document]
Comment (josep-reyero) [reply components/MapFilters.vue:219]: Resolved.

File: components/MapFilters.vue (hunk 1 above)
Comment (josep-reyero) [reply components/MapFilters.vue:227]: Resolved.

File: components/MapFilters.vue (hunk 2 above)
Comment (josep-reyero) [reply components/MapFilters.vue:234]: Resolved.

File: components/MapFilters.vue (hunk 3 above)
Comment (josep-reyero) [reply components/MapFilters.vue:244]: Resolved.

File: components/MapFilters.vue (hunk 5 above)
Comment (josep-reyero) [reply components/MapFilters.vue:278]: Resolved.

File: components/MapFilters.vue (hunk 6 above)
Comment (josep-reyero) [reply components/MapFilters.vue:288]: Resolved.

File: components/MapFilters.vue (hunk 7 above)
Comment (josep-reyero) [reply components/MapFilters.vue:297]: Resolved.

File: components/MapFilters.vue (hunk 8 above)
Comment (josep-reyero) [reply components/MapFilters.vue:315]: Resolved.

File: components/MapFilters.vue (hunk 9 above)
Comment (josep-reyero) [reply components/MapFilters.vue:333]: Resolved.

File: pages/death.vue (hunk 12 above)
Comment (josep-reyero) [reply pages/death.vue:192]: Done.

File: components/MapFilters.vue (hunk 10 above)
Comment (josep-reyero) [reply components/MapFilters.vue:349]: Resolved.

File: pages/detention.vue [hunk 27]
```diff
@@ -276,14 +231,18 @@
...
-               hasWikiPage: wikiPage === '' ? 'no wiki' : 'has wiki',
+               hasWikiPage: wikiPage === '' ? 'noWiki' : 'hasWiki',
                residences: [],
                placeOfDeath: placeOfDeath,
                placeOfDeathCoords: placeOfDeathCoords,
                mannerOfDeath: mannerOfDeath,
-               detentions: []
+               detentions: [],
+               witchState: {
+                 onOff: true,
```
[the remark this answers is not in the document]
Comment (josep-reyero) [reply pages/detention.vue:243]: Resolved.

File: pages/index.vue [hunk 28]
```diff
@@ -304,6 +259,10 @@
                detentions: [],
                investigationDate: investigationDate,
                year: year,
+               witchState: {
+                 onOff: true,
```
[the remark this answers is not in the document]
Comment (josep-reyero) [reply pages/index.vue:263]: Resolved.

File: pages/death.vue [hunk 29]
```diff
@@ -276,14 +226,18 @@
...
                wikiPage: wikiPage,
                hasWikiPage: wikiPage === '' ? 'noWiki' : 'hasWiki',
                residences: [],
                placeOfDeath: placeOfDeath,
                placeOfDeathCoords: placeOfDeathCoords,
                mannerOfDeath: mannerOfDeath,
-               detentions: []
+               detentions: [],
+               witchState: {
+                 onOff: true,
```
[the remark this answers is not in the document]
Comment (josep-reyero) [reply pages/death.vue:238]: Resolved.

File: components/MapFilters.vue [hunk 30]
```diff
@@ -74,188 +138,287 @@
...
 
-       for (let i = 0, len = witches.length; i < len; i++) {
-         let witch = witches[i]
+         for (let w = 0; w < marker.witches.length; w++) {
+           let witch = marker.witches[w];
 
-         if (witch[this.currentFilterProperty] !== filterType) {
-           filteredWitches.push(witch);
+           if (witch[filterProperty] === filterType) {
+             witch.witchState.onOff = false;
```
[the remark this answers is not in the document]
Comment (josep-reyero) [reply components/MapFilters.vue:218]: Resolved.

File: components/CookieBanner.vue [hunk 31]
```diff
@@ -0,0 +1,68 @@
...
+        v-if="!hideBanner">
+    <p class="text-xs md:text-sm ml-2 mt-1">
+      We use cookies to analyse site traffic to understand our
+      users' behaviour and how the site is doing, and to store
+      your cookie preferences. This includes Google Analytics
+      anonymised data.
+      If you wish to turn off Google Analytics you can do so
+      <a class="normal" href="https://tools.google.com/dlpage/gaoptout">here</a>.
+      If you wish to turn off preference cookies click Reject preference cookies.
+      By cliking Accept all you consent to all cookies and Google Analytics
```
Comment (richeduni) [inline components/CookieBanner.vue:13]: Small typo. Should be clicking and not cliking

**Pull request closed without merging**